### DISCUSSION TITLE
Lab6 How to verify my answer from notebook ??

### DISCUSSION THREAD

**Discussion by Russ420** [bol-edu/HLS-SOC-Discussions#175] [Q&A] [answered]
想請問大家，由於 submission guideline裡面有提到需要 "verify your answer from notebook"

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/56719265/e554be68-35a2-4334-af1a-0ab105d927d0)

但是目前有遭遇到兩個問題：
1. 我們使用 jupyter notebook 去讀取 0x1C 的數值都只能得到 ab60 與 ab61。我們認為 ps 端的掃描對於 caravel 來說太慢，導致 ps 端無法抓取到中間運算的 checkbits。

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/56719265/7560e3a8-2335-4675-a055-254a48f6d477)

2. 而且在輸出 checkbits 時 uart 也無法並行運作，附上程式碼：

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/56719265/124a2bb0-4ca2-4baa-ac52-52d94cbe86a1)

**Comment by stage0831** [comment]
幫推個，我們目前也是在這個部分卡住，我們只使用task的方式加上用while卡住等待收值，如下面兩張圖
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/72055234/95b69dd1-228b-4175-9dc9-efdd3e5442fd)
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/72055234/d904347b-307d-41bf-8ba6-c4d385dcdeb1)
我們的推論也是mprj在firmware執行的過程變化的比起PS端的快。這邊python執行的結果有時候會可以執行全部，但有時候不行，會卡在中間的while裡。

目前不確定是不是要使用到asyncio的套件裡面更深入的功能來做驗證。

**Comment by TKAndyLiang** [comment, marked as answer]
我們這組有思考過這個問題，假如能夠找到一個可以從PS端輸入進行procedure的控制的port的話，就可以手動限制速度並且檢查答案，但我們找了一陣子找不太到。
所以我們用另一種有點偷吃步的方法，直接讓firmware的執行速度變慢，也就是在收到答案的值得間隔中安插while loop讓它空轉，有測試過這樣就可以收到所有你想檢查的值。

**Reply by Russ420** [reply]
感謝大家的回覆！

**Reply by weixiang1118** [reply]
請問delay大概要設多少才可以運行呢? 謝謝

**Reply by JIAJIA4** [reply]
設10000還會有些跑不到，我設了100000就都可以測到

**Comment by boledulab** [comment]
雖然說是偷吃步，是一個方法驗證功能的正確性，這在初期驗證階段可以使用，之後可以拿掉。
